ucum/ucum.go

```diff
@@ -0,0 +1,353 @@
+	if factor, ok := getConversionFactor(fromUnit, toUnit); ok {
+		return fromVal * factor, nil
+	}
+	
+	// If we get here, we couldn't find a conversion path
+	return 0, fmt.Errorf("cannot convert from '%s' to '%s'", fromUnit, toUnit)
+}
+
+// GetProductOfUnits returns the product of two units (for multiplication)
+func GetProductOfUnits(unit1, unit2 string) string {
+	unit1, unit2 = FixEmptyUnit(unit1), FixEmptyUnit(unit2)
+	
+	// Handle special cases
+	if unit1 == "1" {
+		return unit2
+	}
+	if unit2 == "1" {
+		return unit1
+	}
+	
+	// Handle unit squaring (when the same unit is multiplied by itself)
+	if unit1 == unit2 {
+		// Return the unit with a squared notation (unit2)
+		return fmt.Sprintf("%s2", unit1)
+	}
+	
+	// Handle special unit combinations based on UCUM rules
+	// TODO: Add more special cases as needed
+	
+	// For simple units, concatenate with a dot
+	return fmt.Sprintf("%s.%s", unit1, unit2)
+}
+
+// GetQuotientOfUnits returns the quotient of two units (for division)
+func GetQuotientOfUnits(unit1, unit2 string) string {
+	unit1, unit2 = FixEmptyUnit(unit1), FixEmptyUnit(unit2)
+	
+	// Handle special cases
+	if unit1 == unit2 {
+		return "1"  // Same units cancel out
+	}
+	if unit2 == "1" {
+		return unit1 // Dividing by 1 gives original unit
+	}
+	
+	// For simple units, we can represent with a slash
+	// This is a simplified implementation
+	return fmt.Sprintf("%s/%s", unit1, unit2)
+}
+
+// validateUCUMSyntax provides basic validation of UCUM syntax
+func validateUCUMSyntax(unit string) bool {
+	// Handle empty unit case
+	if unit == "" || unit == "1" {
+		return true
+	}
+	
+	// Check if it's a known unit from our conversion tables
+	for baseUnit, factors := range commonUnitFactors {
+		if unit == baseUnit {
+			return true
+		}
+		for derivedUnit := range factors {
+			if unit == derivedUnit {
+				return true
+			}
+		}
+	}
+	
+	// Check if it's a known CQL date unit in UCUM format
+	for _, ucumUnit := range CQLToUCUMDateUnits {
+		if unit == ucumUnit {
+			return true
+		}
+	}
+	
+	// Basic syntax check for common patterns
+	if strings.Contains(unit, "/") {
+		// Handle division format (e.g., "m/s")
+		parts := strings.Split(unit, "/")
+		if len(parts) == 2 {
+			return validateUCUMSyntax(parts[0]) && validateUCUMSyntax(parts[1])
+		}
+	}
+	
+	if strings.Contains(unit, ".") {
+		// Handle multiplication format (e.g., "N.m")
+		parts := strings.Split(unit, ".")
+		for _, part := range parts {
+			if !validateUCUMSyntax(part) {
+				return false
+			}
+		}
+		return true
+	}
+	
+	// Check for common UCUM suffixes like -2, 2, etc.
+	// This is a very simplified check
+	if len(unit) > 1 {
+		lastChar := unit[len(unit)-1]
+		if lastChar >= '0' && lastChar <= '9' {
+			baseUnit := unit[:len(unit)-1]
+			return validateUCUMSyntax(baseUnit)
+		}
+	}
+	
+	// Default to accepting for now to avoid false negatives
+	// A more comprehensive implementation would check against a full UCUM database
+	return true
```

Comment:
I'm not sure I agree with this, I think we should return an error here if we run across a case that we can't yet handle and provide enough detailed information that a user can determine if the unit being processed is invalid on their side, or a case where we have a bug, i.e. we shouldn't be suppressing errors that our engine actually has to avoid false negatives/positives.